includes/stores/multistore.php

```diff
@@ -0,0 +1,256 @@
+<?php
+
+/**
+ * Token storage for a Multi-Site install. Works with multiple users, sharing
+ * connections across multiple sites.
+ *
+ * Stores the tokens in CPTs, and the shared status as postmeta on those posts.
+ *
+ * @package Keyring
+ */
+class Keyring_MultiStore extends Keyring_Store {
+
+	/**
+	 * Initialises the store.
+	 *
+	 * @return Keyring_MultiStore The store instance.
+	 */
+	static function &init() {
+		static $instance = false;
+
+		if ( ! $instance ) {
+			register_post_type(
+				'kr_request_token',
+				array(
+					'label'       => __( 'Keyring Request Token', 'keyring' ),
+					'description' => __( 'Token or authentication details stored by Keyring. Request tokens are used during the authorization flow.', 'keyring' ),
+					'public'      => false,
+				)
+			);
+
+			register_post_type(
+				'kr_access_token',
+				array(
+					'label'       => __( 'Keyring Access Token', 'keyring' ),
+					'description' => __( 'Token or authentication details stored by Keyring. Access tokens are used to make secure requests.', 'keyring' ),
+					'public'      => false,
+				)
+			);
+
+			$instance = new Keyring_MultiStore;
+		}
+
+		return $instance;
+	}
+
+	/**
+	 * Inserts a new token in the store.
+	 *
+	 * @param Keyring_Token $token  The token being stored.
+	 * @return int|false The unique ID assigned to the token, or false if it fails to save.
+	 */
+	function insert( $token ) {
+		$site_id = get_current_blog_id();
+		switch_to_blog( KEYRING__MULTI_STORE_SITE_ID );
+
+		// Avoid duplicates by checking to see if this exists already
+		$found = get_posts(
+			array(
+				'posts_per_page' => 1,
+				'post_type'      => 'kr_' . $token->type() . '_token',
+				'meta_key'       => 'service',
+				'meta_value'     => $token->get_name(),
+				'author'         => get_current_user_id(),
+				's'              => serialize( $token->token ), // Search the post content for this token
+				'exact'          => true, // Require exact content match
+				'sentence'       => true, // Require to search by phrase, otherwise string is split by regex
+			)
+		);
+
+		if ( $found ) {
+			$token->unique_id = $found[0]->ID;
+			restore_current_blog();
+			return $this->update( $token, $shared );
```

Comment:
Does `$shared` have a value at this point?

Reply:
Ugh, I forgot to rework that. `$shared` was passed as a parameter in an earlier version, but that doesn't work inheriting from `Keyring_Store`.